components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/handlers/DefaultKeyValidationHandler.java

```diff
@@ -170,32 +180,66 @@ public boolean validateScopes(TokenValidationContext validationContext) throws A
                 + ":" +
                 validationContext.getHttpVerb();
 
-        boolean isValid = false;
+        Set<OAuth2ScopeValidator> oAuth2ScopeValidators = OAuthServerConfiguration.getInstance()
+                .getOAuth2ScopeValidators();
+        //validate scope for filtered validators from db
+        String[] scopeValidators;
+        OAuthAppDO appInfo;
         try {
-            Set<OAuth2ScopeValidator> oAuth2ScopeValidators = OAuthServerConfiguration.getInstance()
-                    .getOAuth2ScopeValidators();
-            //validate scope from all the validators
+            OAuthAppDAO oAuthAppDAO = new OAuthAppDAO();
+            appInfo = oAuthAppDAO.getAppInformation(clientId);
+            scopeValidators = appInfo.getScopeValidators();     //get scope validators from the DB
+            List<String> appScopeValidators=  new ArrayList<>(Arrays.asList(scopeValidators));
+
+            if (ArrayUtils.isEmpty(scopeValidators)) {
+                log.debug(String.format("There is no scope validator registered for %s@%s", appInfo.getApplicationName(),
+                            OAuth2Util.getTenantDomainOfOauthApp(appInfo)));
+                return true;
+            }
+
             for (OAuth2ScopeValidator validator : oAuth2ScopeValidators) {
-                if (validator != null) {
-                    if (validator.validateScope(accessTokenDO, resource)) {
-                        isValid = true;
-                    } else {
-                        if (log.isDebugEnabled()) {
-                            log.debug("Scope validation failed from "+ validator);
+                try {
+                    if (validator != null && appScopeValidators.contains(validator.getValidatorName())) {
+                        //take the intersection of defined scope validators and scope
+                        //validators registered for the application
+                        log.debug(String.format("Validating scope of token %s using %s", accessTokenDO.getTokenId(),
+                                validator.getValidatorName()));
+
+                        boolean isValid = validator.validateScope(accessTokenDO, resource);
+                        if (!isValid) {
+                            log.debug(String.format("Scope validation failed for %s", validator.getValidatorName()));
```

Comment:
Please add some details from accesstokenDO to identify what scopes sent and why it failed. Reason is this is not really issue with validator but problem with data. So we can easily identify them.

Reply:
Fixed with b58b5414eb4051494cf7768e5ebfea2d5d7a5933